examples/gno.land/p/demo/avl/pager/pager.gno

```diff
@@ -0,0 +1,220 @@
+
+	items := []Item{}
+	p.Tree.ReverseIterateByOffset(startIndex, endIndex-startIndex, func(key string, value interface{}) bool {
+		items = append(items, Item{Key: key, Value: value})
+		return false
+	})
+
+	page.Items = items
+	page.PageNumber = pageNumber
+	page.HasPrev = pageNumber > 1
+	page.HasNext = pageNumber < totalPages
+	return page
+}
+
+func (p *Pager) MustGetPageByPath(rawURL string) *Page {
+	page, err := p.GetPageByPath(rawURL)
+	if err != nil {
+		panic("invalid path")
+	}
+	return page
+}
+
+// GetPageByPath retrieves a page of results based on the query parameters in the URL path.
+func (p *Pager) GetPageByPath(rawURL string) (*Page, error) {
+	pageNumber, pageSize, err := p.ParseQuery(rawURL)
+	if err != nil {
+		return nil, err
+	}
+	return p.GetPageWithSize(pageNumber, pageSize), nil
+}
+
+// UI generates the Markdown UI for the page selector.
+func (p *Page) Selector() string {
+	pageNumber := p.PageNumber
+	pageNumber = max(pageNumber, 1)
+
+	if p.TotalPages <= 1 {
+		return ""
+	}
+
+	md := ""
+
+	if p.HasPrev {
+		// Always show the first page link
+		md += ufmt.Sprintf("[%d](?%s=%d) | ", 1, p.Pager.PageQueryParam, 1)
+
+		// Before
+		if p.PageNumber > 4 {
+			md += "… | "
+		}
+
+		if p.PageNumber > 3 {
+			md += ufmt.Sprintf("[%d](?%s=%d) | ", p.PageNumber-2, p.Pager.PageQueryParam, p.PageNumber-2)
+		}
+
+		if p.PageNumber > 2 {
+			md += ufmt.Sprintf("[%d](?%s=%d) | ", p.PageNumber-1, p.Pager.PageQueryParam, p.PageNumber-1)
+		}
+	}
+
+	if p.PageNumber > 0 && p.PageNumber <= p.TotalPages {
+		// Current page
+		md += ufmt.Sprintf("**%d**", p.PageNumber)
+	} else {
+		md += ufmt.Sprintf("_%d_", p.PageNumber)
+	}
+
+	if p.HasNext {
+		md += " | "
+
+		if p.PageNumber < p.TotalPages-1 {
+			md += ufmt.Sprintf("[%d](?%s=%d) | ", p.PageNumber+1, p.Pager.PageQueryParam, p.PageNumber+1)
+		}
+
+		if p.PageNumber < p.TotalPages-2 {
+			md += ufmt.Sprintf("[%d](?%s=%d) | ", p.PageNumber+2, p.Pager.PageQueryParam, p.PageNumber+2)
+		}
+
+		if p.PageNumber < p.TotalPages-3 {
+			md += "… | "
+		}
+
+		// Always show the last page link
+		md += ufmt.Sprintf("[%d](?%s=%d)", p.TotalPages, p.Pager.PageQueryParam, p.TotalPages)
+	}
+
+	return md
+}
+
+// ParseQuery parses the URL to extract the page number and page size.
+func (p *Pager) ParseQuery(rawURL string) (int, int, error) {
+	u, err := url.Parse(rawURL)
+	if err != nil {
+		return 1, p.DefaultPageSize, err
+	}
+
+	query := u.Query()
+	pageNumber := 1
+	pageSize := p.DefaultPageSize
+
+	if p.PageQueryParam != "" {
+		if pageStr := query.Get(p.PageQueryParam); pageStr != "" {
+			pageNumber, err = strconv.Atoi(pageStr)
+			if err != nil || pageNumber < 1 {
+				pageNumber = 1
+			}
+		}
+	}
+
+	if p.SizeQueryParam != "" {
+		if sizeStr := query.Get(p.SizeQueryParam); sizeStr != "" {
+			pageSize, err = strconv.Atoi(sizeStr)
+			if err != nil || pageSize < 1 {
+				pageSize = p.DefaultPageSize
+			}
+		}
+	}
+
+	return pageNumber, pageSize, nil
+}
+
+func min(a, b int) int {
```

Comment:
seems not used in any place

Reply:
[`3754a70` (#2584)](https://github.com/gnolang/gno/pull/2584/commits/3754a7093a6f658d6be474340f03e0cd44964c34)